### PR TITLE
feat(resource-types): add resource-types

### PR DESCRIPTION
**Pull request by toumorokoshi**
Porting the resource types AEP from google.aip.dev.

Specific changes:

- Rename from resource name to resource path, to be consistent with
  aeps.
- Instead of introducing new terminology within the AEP, updating
  /glossary instead.

### CODE REVIEW

File: aep/general/0004/aep.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+## Rationale
+
+### Singular and Plural
+
+Well-defined singular and plurals of a resource enable clients to determine the
+proper name to use in code and documentation.
+
+lowerCamelCase can be translated into other common forms of a resource path
+such as UpperCamelCase and snake_case.
```
Comment (rofrankel) [inline aep/general/0004/aep.md:90]: I don't know exactly what this is the rationale for...and all the string cases can be converted to the others, right?

**Review by rofrankel — CHANGES_REQUESTED** [review]

File: aep/general/0004/aep.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    pattern: "projects/{project}/topics/{topic}"
+    singular: "topic"
+    plural: "topics"
+  };
+
+  // Name and other fields...
+}
+```
+
+- Patterns **must** correspond to the [resource path][resource-paths].
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:55]: resource paths are the spec, correct? If we define a spec, I'd suggest it is defined in resource path and we link to there.

File: aep/general/0004/aep.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+- Pattern variables (the segments within braces) **must** use `snake_case`, and
+  **must not** use an `_id` suffix.
+- Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:58]: That's not invalid, I believe - In practice it maps to a protobuf field name, and I think a_____9 is valid?

File: aep/general/0004/aep.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+- Pattern variables **must** be unique within any given pattern. (e.g.
+  `projects/{abc}/topics/{abc}` is invalid; this is usually a natural corollary
+  of collection identifiers being unique within a pattern.)
+- Singular **must** be the lower camel case of the type.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:62]: Singular variable - I'll update to clarify. 

TLDR on the snake / camel discrepancy - yeah they should match, probably snake, but:

There is a distinction between pattern variables and singular types - they don't necessarily have to match to work: pattern variables match to proto fields in Google's API frameworks, while the singular / plural is used more to match with collection names. Of course, with the advent of shortening the collection name, we can't even rely on singular / plural being used in the path at all anymore.

We should normalize it, but I think it'd be good to clarify what sort of consistency users should expect based on this - with resource path singletons, API paths, and no guarantee that path segments match collection names, there's basically no technical reason for any of this to match and clients have to assume none of them do.

File: aep/general/0004/aep.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  - Pattern variables **must** be the singular form of the resource type e.g. a
+    pattern variable representing a `Topic` resource ID is named `{topic}`.
+- Plural **must** be the lower camel case plural of the singular.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:65]: value in the annotation - I'll update.

File: aep/general/0004/aep.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  - Pattern collection identifier segments **must** match the plural of the
+    resources, except in the case of [nested collections][].
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:67]: ack, you're right.

File: aep/general/0004/aep.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+#### Pattern uniqueness
+
+When multiple patterns are defined within a resource, these patterns **must**
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:71]: I think it's a good suggestion! added.

File: aep/general/0004/aep.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:55]: Well...what I meant is that the pattern uses curly-brace variables that aren't defined by the resource-paths AEP.

File: aep/general/0004/aep.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:58]: I agree that's a valid field name, but pattern variables don't correspond to a field name, they correspond to a the name of a resource type (which should be in PascalCase, and converting PascalCase to snake_case never results in adjacent underscores).

File: aep/general/0004/aep.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:62]: Are you thinking of google.api.http routes (e.g. `{parent=...}` rather than google.api.resource patterns?

File: aep/general/0004/aep.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:71]: Did this ever get updated?

File: aep/general/0004/aep.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type}`. The type name:
+
+- **must** Start with an uppercase letter.
+- **must** Only contain alphanumeric characters.
+- **must** Be of the singular form of the noun.
+- **must** Use PascalCase (UpperCamelCase).
+- For Kubernetes, the type name **must** match the [object][] name.
+- For Protobuf, the type name **must** match the name of the protobuf message.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:24]: The guidance below talks about the usage of a proto annotation, which doesn't necessarily have to match. So I think it's good to be explicit here.

Added OAS guidance.

File: aep/general/0004/aep.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+- `pubsub.googleapis.com/Topic`
+- `pubsub.googleapis.com/Subscription`
+- `spanner.googleapis.com/Database`
+- `spanner.googleapis.com/Instance`
+- `networking.istio.io/Instance`
+
+### Annotating resource types
+
+APIs **must** annotate the resource types for each resource in the API using
+the [`google.api.resource`][resource] annotation:
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:39]: I'd rather focus on proto for now and see how things like in OAS later at the end - primarily because I have the aepc repository which just uses what grpc-proxy translates things as.

How about an issue on it for later?

File: aep/general/0004/aep.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+### Examples
+
+Examples of resource types include:
+
+- `pubsub.googleapis.com/Topic`
+- `pubsub.googleapis.com/Subscription`
+- `spanner.googleapis.com/Database`
+- `spanner.googleapis.com/Instance`
+- `networking.istio.io/Instance`
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:34]: done.

File: aep/general/0004/aep.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:62]: I'm referring to the section above.

File: aep/general/0004/aep.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  - Pattern variables **must** be the singular form of the resource type e.g. a
+    pattern variable representing a `Topic` resource ID is named `{topic}`.
+- Plural **must** be the lower camel case plural of the singular.
+  - Pattern collection identifier segments **must** match the plural of the
+    resources, except in the case of [nested collections][].
+
+#### Pattern uniqueness
+
+When multiple patterns are defined within a resource, these patterns **must**
+be mutually unique, where uniqueness is defined as being by-character identical
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:72]: nope, removed.

File: aep/general/0004/aep.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+once all resource ID path segments have been removed, leaving all `/`
+separators.
+
+Therefore the following two patterns **must not** be defined within the same
+resource:
+
+- `user/{user}`
+- `user/{user_part_1}~{user_part_2}`
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:80]: I think I get your point - the paragraph is kind of roundabout?

I tried to focus specifically on the fact that it has to ensure unique matches - PTAL.

File: aep/general/0004/aep.md (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0004/aep.md:90]: I think the main issue was acronyms - e.g. if you have a resource name APIMethod for example - how would one know that should be translated to API_method?

Although I guess should we just have the type be in kebab-case, to match the resource path?

File: aep/general/0004/aep.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:58]: corresponding to a resource type seems reasonable - it may also avoid the issue where it's hard to figure our the parent resource types in the pattern, since the collection names can be short-hand for the resource (that clause in the resource path about https://aep.dev/122#nested-collections)

In that case we would have stuff like:

```/users/{user}/events/{userEvent}```

is that ok with you?

File: aep/general/0004/aep.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+use `Instance` as a type name.
+
+When mapping the relationships between APIs and their resources, however, it
+becomes important to have a single, globally-unique type name. Additionally,
+tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
+
+## Guidance
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type}`. The type name:
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:17]: Do you have a suggestion? I at least called it "Type name" to not feel as recursive.

File: aep/general/0004/aep.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: yes, I think you may be looking at an outdated diff. Although I rewrote this sentence anyway.

File: aep/general/0004/aep.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:57]: ack, yeah I can define it. Let's resolve https://github.com/aep-dev/aep.dev/pull/76/files#r1398519206 then I'll add it.

File: aep/general/0004/aep.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:60]: Actually I was thinking through this a little more - and I think that in general, patterns a somewhat flawed - they force you to re-enumerate the resource paths of parents, or worse have a parent resource path that doesn't match.

Really the only thing a sub-resource can do is add a suffix to a specific parent. So really, I think we should do something like:

```
google.api.resource = {
    type: "user-event"
    singular: "user-event"
    plural: "user-events"
    parents = [ 
        type: "user",
        path_suffix = "/events/{user-event}"
    ]
}
```

does that make sense? I think we'd have to come up with our own resource annotation to support it, but it allows for generation of correct, fully qualified resource paths.

File: aep/general/0004/aep.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:67]: I don't understand the current phrasing (and also, should it be snake_case rather than lowerCamelCase)?

Separately, elsewhere in this file it's spelled "lowerCamelCase" rather than "lower camel case".

File: aep/general/0004/aep.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  - Pattern variables **must** be the singular form of the resource type e.g. a
+    pattern variable representing a `Topic` resource ID is named `{topic}`.
+- Plural **must** be the lower camel case plural of the singular.
+  - Pattern collection identifier segments **must** match the plural of the
+    resource, except in the case of [nested collections][].
+
+#### Pattern uniqueness
+
+If multiple patterns are defined within a resource, the patterns defined must
+not overlap in the set of resource paths that they can match. In other words, a
```
Comment (rofrankel) [inline aep/general/0004/aep.md:74]:
```suggestion
If multiple patterns are defined within a resource, the patterns defined **must
not** overlap in the set of resource paths that they can match. In other words, a
```

File: aep/general/0004/aep.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+resource path may match at most one of the patterns.
+
+Therefore the following two patterns **must not** be defined within the same
```
Comment (rofrankel) [inline aep/general/0004/aep.md:77]:
```suggestion
For example, the following two patterns would not be valid for the same
```

`**must not**` in an example feels a bit off to me, so suggesting slightly different phrasing.

File: aep/general/0004/aep.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:81]: Seems better, thanks.  I added a couple superficial suggestions on that text, but resolving this comment.

File: aep/general/0004/aep.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:41]: Hmm...we use proto at Roblox so it doesn't affect me in that sense, but I wonder how our OAS-aligned members will feel about OAS being effectively a secondary consideration. :/

File: aep/general/0004/aep.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0004/aep.md:26]: Gotcha.  Not marking this resolved because there's still a casing suggestion that isn't committed.

**Review by rofrankel — CHANGES_REQUESTED** [review]

File: aep/general/0004/aep.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+becomes important to have a single, globally-unique type name. Additionally,
+tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
+
+## Guidance
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type Name}`. The type name:
+
+- **must** Start with an uppercase letter.
+- **must** Only contain alphanumeric characters.
```
Comment (gibson042) [inline aep/general/0004/aep.md:20]:
```suggestion
- **must** Only contain ASCII alphanumeric characters.
```

File: aep/general/0004/aep.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:41]: I'll leave it open, we can raise concerns there. Personally I'd like to see if we can find a better way to have dual support for OAS / proto than spelling out every annotation on each (although maybe that's the way it has to be).

File: aep/general/0004/aep.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:60]: the pattern annotation is used by the api-linter, but it's not used for gRPC-HTTP bindings at all.

My opinion is that we should fork the annotation here so it's a bit more semantically clear, and work from parent references. But @rofrankel noted in our last meeting that having all patterns enumerated is nice from a documentation perspective.

We've agreed upon keeping the existing annotation as-is, but introducing a more standard resource definition syntax into the AEPs.

I might try to just fix this AEP to work without it though (e.g. documenting the pattern syntax as suggested) so that we can get this pretty critical AEP in while we debate resource syntax.

File: aep/general/0004/aep.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: added type name, PTAL.

File: aep/general/0004/aep.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: okay, I rewrote this section, PTAL. added a grammar and more specific guidance, let's talk there.

File: aep/general/0004/aep.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md:65]: so after some thinking, I've updated this section to use kebab-case to be consistent with things like resource path. 

Why do you think it should be snake_case? I was thinking kebab-case to remove yet another place where there's a mostly similar string that has to be transformed somehow.

File: aep/general/0004/aep.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: added OAS annotations.

File: aep/general/0004/aep.md (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: updated to kebab-case.

File: aep/general/0004/aep.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: updated the AEP quite a bit to be clearer here, including an grammar. Please review and leave any comments on the new text.

File: aep/general/0004/aep.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0004/aep.md]: fixed the aep as-is for now, I'll work on adding the resource definition at a future time.

File: aep/general/0004/aep.md [hunk 17]
```diff
@@ -16,65 +16,99 @@ tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
...
-- `pubsub.example.com/Subscription`
-- `spanner.example.com/Database`
-- `spanner.example.com/Instance`
-- `networking.istio.io/Instance`
+- `networking.istio.io/instance`
+- `pubsub.example.com/topic`
+- `pubsub.example.com/subscription`
+- `spanner.example.com/database`
+- `spanner.example.com/instance`
+- `user.example.com/user-event`
```
Comment (rofrankel) [inline aep/general/0004/aep.md:39]: What about the `apis.example.com/users/user-event` form?  I don't want to argue for that to be _the_ canonical form, but I do think it should be supported.  Otherwise we are prescribing that everyone has to use a subdomain for each API, which raises the barrier to entry (and would have been a dealbreaker for Roblox, for example).

File: aep/general/0004/aep.md [hunk 18]
```diff
@@ -16,65 +16,99 @@ tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
...
+```
+
+{% tab oas %}
+
+For OpenAPI 3.0, use the `x-aep-resource` extension:
+
+```json
+{
+  "type": "object",
+  "x-aep-resource": {
```
Comment (rofrankel) [inline aep/general/0004/aep.md:72]: Is this extension defined anywhere e.g. as a JSON Schema?  (Does it need to be?)

File: aep/general/0004/aep.md [hunk 19]
```diff
@@ -16,65 +16,99 @@ tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
...
-  of collection identifiers being unique within a pattern.)
-- Singular **must** be the lower camel case of the type.
-  - Pattern variables **must** be the singular form of the resource type e.g. a
-    pattern variable representing a `Topic` resource ID is named `{topic}`.
-- Plural **must** be the lower camel case plural of the singular.
-  - Pattern collection identifier segments **must** match the plural of the
-    resource, except in the case of [nested collections][].
+{% endtabs %}
+
+- The `singular` field **must** match the type name.
```
Comment (rofrankel) [inline aep/general/0004/aep.md:86]:
```suggestion
- The `singular` field **must** be the kebab-case singular type name.
```

File: aep/general/0004/aep.md [hunk 20]
```diff
@@ -87,8 +121,10 @@ resource:
 Well-defined singular and plurals of a resource enable clients to determine the
 proper name to use in code and documentation.
 
-lowerCamelCase was chosen as the formatting as it can easily translated into
-other common forms of a resource path such as UpperCamelCase and snake_case.
+google.aip.dev defines the resource type to use upper camel case, while aeps
```
Comment (rofrankel) [inline aep/general/0004/aep.md:124]:
```suggestion
google.aip.dev defines the resource type to use UpperCamelCase, while aeps
```

File: aep/general/0004/aep.md [hunk 21]
```diff
@@ -87,8 +121,10 @@ resource:
...
+define the kebab-case. This is to enforce better consistency in the
```
Comment (rofrankel) [inline aep/general/0004/aep.md:125]:
```suggestion
google.aip.dev uses upper camel case for resource types, while aep.dev
uses kebab-case. This is to enforce better consistency in the
```

File: aep/general/0003/aep.md [hunk 22]
```diff
@@ -68,8 +68,7 @@ An object upon which one or more methods can operate.
 
 ### API Resource Type
 
-The type of an object exposed via an API. It is globally unique within an
-API.
+The type of an object exposed via an API. It is globally unique within an API.
```
Comment (rofrankel) [inline aep/general/0003/aep.md:71]: Is "object" the right word here?  An API resource isn't exactly the same thing as an "object" in OOP terms, even though there are some parallels, so this might be confusing in that sense.  Maybe "entity"?

**Review by rofrankel — CHANGES_REQUESTED** [review]

File: aep/general/0004/aep.md (hunk 18 above)
Comment (toumorokoshi) [reply aep/general/0004/aep.md:72]: it's a good question! It definitely should be documented somewhere.

The question is where. I'm thinking maybe one of:

1. an AEP
2. in something like your aep repository


Personally I err toward something a little more fully encapsulated, and think we should document it in this repo. But I'm flexible. Definitely a topic we can discuss as a group?

File: aep/general/0004/aep.md (hunk 17 above)
Comment (toumorokoshi) [reply aep/general/0004/aep.md:39]: We do have an outstanding issue to pick up this discussion at some point: https://github.com/aep-dev/aep.dev/issues/74#issue-2000272320.

I'll update the example as suggested, but my main concern is we have to go put these types of examples everywhere - ideally we'd consolidate discussion of a single concept to a single AEP so we don't have to go rewrite a ton of examples as things change.

File: aep/general/0003/aep.md (hunk 22 above)
Comment (toumorokoshi) [reply aep/general/0003/aep.md:71]: I just realized that we defined "API resource" above, so I'll just re-use that.

FWIW I used the term object to match Kubernetes: https://kubernetes.io/docs/concepts/overview/working-with-objects/. Terraform calls these resources, so I think we're already in a world where there's duplicate terminology.

But entity is as good as any - I updated the API resource above to use that term, and updated this section to use the term "API resource".

File: aep/general/0004/aep.md (hunk 18 above)
Comment (toumorokoshi) [reply aep/general/0004/aep.md:72]: schema added, ptal.

**Review by rofrankel — APPROVED** [review]

**Comment by toumorokoshi** [comment]
@earth2marsh I'm going to go ahead and merge this one to get it off the list. Feel free to bring up any issues with the json schema and I'm happy to address them in a follow-up.

File: aep/general/0004/aep.md (hunk 18 above)
Comment (earth2marsh) [reply aep/general/0004/aep.md:72]: I'm not sure I understand the goal here from the snippets of conversation. It might help me to have an explanation of what `x-aep-resource` is meant to capture? I suspect there's a limitation in the OpenAPI that you're trying to compensate for?

**Comment by earth2marsh** [comment]
sgtm. we can pick up my comment above (must have been writing it at the same time) independently.
